File: filter-service/app/oklab.py
"""
OKLab color conversion + nearest-palette match — server side.

This is the **shared color contract** for trace filters: a cell's mean
colour is converted to OKLab and matched to the nearest palette chip. The
transform is byte-identical to color-lab's `rgb_to_oklab` (the tool that
computed the `lab_munsell` / `lab_grays` OKLab columns in the DB), so a
cell's OKLab lives in the same space as the chips it is matched against.
Transform per Björn Ottosson (2020); color-lab validates the matrices
against `colour.XYZ_to_Oklab`.

A `lib/color/oklab.ts` mirror runs in the client preview; an algorithm
parity test keeps the two in lockstep.
"""
from __future__ import annotations

import numpy as np

# Forward matrices (Ottosson). Inverses are not needed server-side (we only
# go RGB → OKLab). Kept byte-identical to color-lab/build_palette.py.
_OKLAB_M1 = np.array(
    [
        [0.4122214708, 0.5363325363, 0.0514459929],
        [0.2119034982, 0.6806995451, 0.1073969566],
        [0.0883024619, 0.2817188376, 0.6299787005],
    ]
)
_OKLAB_M2 = np.array(
    [
        [0.2104542553, 0.7936177850, -0.0040720468],
        [1.9779984951, -2.4285922050, 0.4505937099],
        [0.0259040371, 0.7827717662, -0.8086757660],
    ]
)


def _srgb_to_linear(c: np.ndarray) -> np.ndarray:
    return np.where(c <= 0.04045, c / 12.92, ((np.maximum(c, 0.0) + 0.055) / 1.055) ** 2.4)


def rgb_to_oklab(rgb01: np.ndarray) -> np.ndarray:
    """sRGB (gamma-encoded, 0..1, shape (..., 3)) → OKLab (..., 3)."""
    lin = _srgb_to_linear(np.asarray(rgb01, dtype=np.float64))
    # NumPy's SIMD matmul path spuriously reports a stale FP-error flag on
    # large operands ("divide by zero in matmul" — impossible, matmul only
    # multiply-adds). Inputs here are finite and in-gamut, so the warnings are
    # pure noise that spams Cloud Run's stderr. Silencing them leaves the
    # output bit-identical (OKLab parity with the TS mirror is unaffected).
    with np.errstate(divide="ignore", over="ignore", invalid="ignore"):
        lms_ = np.cbrt(lin @ _OKLAB_M1.T)
        return lms_ @ _OKLAB_M2.T


def rgb255_to_oklab(rgb255: np.ndarray) -> np.ndarray:
    """sRGB uint8 (0..255, shape (..., 3)) → OKLab (..., 3)."""
    return rgb_to_oklab(np.asarray(rgb255, dtype=np.float64) / 255.0)


def adjust_oklab(
    oklab: np.ndarray,
    hue_deg: float = 0.0,
    lightness_delta: float = 0.0,
    chroma_scale: float = 1.0,
) -> np.ndarray:
    """Apply an OKLCh adjustment to OKLab colours: rotate hue by `hue_deg`,
    scale chroma by `chroma_scale`, shift lightness by `lightness_delta`
    (L clamped to [0, 1]). Shape (..., 3) → (..., 3). Mirror of
    `lib/color/oklab.ts::adjustOklab`.

    Used by Circulate's inner ellipse: the cell mean is adjusted by the chosen
    sub colour filter, then snapped back to the nearest palette chip
    (`nearest_palette_indices`) so the result never leaves the palette. The
    identity adjustment `(0, 0, 1)` collapses the inner colour to the same chip
    as the outer cell.
    """
    lab = np.asarray(oklab, dtype=np.float64)
    a = lab[..., 1]
    b = lab[..., 2]
    chroma = np.hypot(a, b) * chroma_scale
    hue = np.arctan2(b, a) + np.radians(hue_deg)
    out = np.empty_like(lab)
    out[..., 0] = np.clip(lab[..., 0] + lightness_delta, 0.0, 1.0)
    out[..., 1] = chroma * np.cos(hue)
    out[..., 2] = chroma * np.sin(hue)
    return out


def nearest_palette_indices(cell_oklab: np.ndarray, palette_oklab: np.ndarray) -> np.ndarray:
    """For each cell OKLab (N, 3), the index of the nearest palette chip
    (M, 3) by squared euclidean OKLab distance. Returns (N,) int indices.

    Plain numpy broadcasting — no scipy/cKDTree dependency. M ≤ 128 and N is
    the cell count, so the (N, M) distance matrix is small and fast.
    """
    cells = np.asarray(cell_oklab, dtype=np.float64).reshape(-1, 3)
    palette = np.asarray(palette_oklab, dtype=np.float64).reshape(-1, 3)
    # (N, 1, 3) - (1, M, 3) → (N, M, 3) → sum sq over axis 2 → (N, M)
    d2 = ((cells[:, None, :] - palette[None, :, :]) ** 2).sum(axis=2)
    return d2.argmin(axis=1)

File: lib/color/oklab.ts
/**
 * OKLab color conversion + nearest-palette match — client side.
 *
 * Mirror of `filter-service/app/oklab.py` (and color-lab's `rgb_to_oklab`,
 * which computed the `lab_munsell` / `lab_grays` OKLab columns in the DB).
 * The matrices + gamma are byte-identical so a cell's OKLab lives in the
 * same space as the palette chips it is matched against. Transform per
 * Björn Ottosson (2020).
 *
 * Used by the trace live-preview (`trace-cell-colors.ts`); an algorithm
 * parity test (`oklab.test.ts` + `test_oklab.py`) keeps client and server
 * in lockstep against shared reference vectors.
 */

/** OKLab triple `[L, a, b]`. */
export type Oklab = readonly [number, number, number]

// Forward matrices (Ottosson), identical to the Python/color-lab source.
const M1 = [
  [0.4122214708, 0.5363325363, 0.0514459929],
  [0.2119034982, 0.6806995451, 0.1073969566],
  [0.0883024619, 0.2817188376, 0.6299787005],
] as const
const M2 = [
  [0.2104542553, 0.7936177850, -0.0040720468],
  [1.9779984951, -2.4285922050, 0.4505937099],
  [0.0259040371, 0.7827717662, -0.8086757660],
] as const

function srgbToLinear(c: number): number {
  return c <= 0.04045 ? c / 12.92 : ((Math.max(c, 0) + 0.055) / 1.055) ** 2.4
}

/**
 * sRGB (gamma-encoded, 0..255) → OKLab. Matches the server byte-for-byte
 * (same matrices, same gamma, `Math.cbrt`).
 */
export function rgb255ToOklab(r: number, g: number, b: number): Oklab {
  const lr = srgbToLinear(r / 255)
  const lg = srgbToLinear(g / 255)
  const lb = srgbToLinear(b / 255)
  const l = Math.cbrt(M1[0][0] * lr + M1[0][1] * lg + M1[0][2] * lb)
  const m = Math.cbrt(M1[1][0] * lr + M1[1][1] * lg + M1[1][2] * lb)
  const s = Math.cbrt(M1[2][0] * lr + M1[2][1] * lg + M1[2][2] * lb)
  return [
    M2[0][0] * l + M2[0][1] * m + M2[0][2] * s,
    M2[1][0] * l + M2[1][1] * m + M2[1][2] * s,
    M2[2][0] * l + M2[2][1] * m + M2[2][2] * s,
  ]
}

/**
 * Rotate the hue of an OKLab colour by `degrees` (OKLCh hue rotation),
 * keeping lightness L and chroma constant. Mirror of `oklab.py`'s
 * `rotate_hue`; used by Circulate's inner ellipse (the cell mean's hue is
 * shifted, then snapped back to the nearest palette chip so it never leaves
 * the palette). `degrees === 0` is the identity.
 */
export function rotateHueOklab(lab: Oklab, degrees: number): Oklab {
  const a = lab[1]
  const b = lab[2]
  const chroma = Math.hypot(a, b)
  const hue = Math.atan2(b, a) + (degrees * Math.PI) / 180
  return [lab[0], chroma * Math.cos(hue), chroma * Math.sin(hue)]
}

/**
 * Index of the nearest palette chip to `lab` by squared euclidean OKLab
 * distance. `palette` must be non-empty; returns 0 for an empty palette.
 */
export function nearestPaletteIndex(lab: Oklab, palette: readonly Oklab[]): number {
  let best = 0
  let bestD = Infinity
  for (let i = 0; i < palette.length; i += 1) {
    const p = palette[i]
    const dl = lab[0] - p[0]
    const da = lab[1] - p[1]
    const db = lab[2] - p[2]
    const d = dl * dl + da * da + db * db
    if (d < bestD) {
      bestD = d
      best = i
    }
  }
  return best
}
